cpp/src/arrow/flight/sql/odbc/odbc_api.cc

```diff
@@ -121,6 +123,108 @@ SQLRETURN SQLFreeHandle(SQLSMALLINT type, SQLHANDLE handle) {
   return SQL_ERROR;
 }
 
+SQLRETURN SQLGetDiagFieldW(SQLSMALLINT handleType, SQLHANDLE handle,
+                           SQLSMALLINT recNumber, SQLSMALLINT diagIdentifier,
+                           SQLPOINTER diagInfoPtr, SQLSMALLINT bufferLength,
+                           SQLSMALLINT* stringLengthPtr) {
+  using driver::odbcabstraction::Diagnostics;
+  using ODBC::GetStringAttribute;
+  using ODBC::ODBCConnection;
+  using ODBC::ODBCEnvironment;
+
+  if (!handle) {
+    return SQL_INVALID_HANDLE;
+  }
+
+  if (!diagInfoPtr) {
+    return SQL_ERROR;
+  }
+
+  // Set character type to be Unicode by defualt (not Ansi)
+  bool isUnicode = true;
+  Diagnostics* diagnostics = nullptr;
+
+  switch (handleType) {
+    case SQL_HANDLE_ENV: {
+      ODBCEnvironment* environment = reinterpret_cast<ODBCEnvironment*>(handle);
+      diagnostics = &environment->GetDiagnostics();
+      break;
+    }
+
+    case SQL_HANDLE_DBC: {
+      ODBCConnection* connection = reinterpret_cast<ODBCConnection*>(handle);
+      diagnostics = &connection->GetDiagnostics();
+      break;
+    }
+
+    default:
+      return SQL_ERROR;
+  }
+
+  if (!diagnostics) {
+    return SQL_ERROR;
+  }
+
+  // Retrieve header level diagnostics if Record 0 specified
+  if (recNumber == 0) {
+    switch (diagIdentifier) {
+      case SQL_DIAG_NUMBER: {
+        SQLINTEGER count = static_cast<SQLINTEGER>(diagnostics->GetRecordCount());
+        *static_cast<SQLINTEGER*>(diagInfoPtr) = count;
+        if (stringLengthPtr) {
+          *stringLengthPtr = sizeof(SQLINTEGER);
+        }
+
+        return SQL_SUCCESS;
+      }
+
+      case SQL_DIAG_SERVER_NAME: {
+        const std::string source = diagnostics->GetDataSourceComponent();
```

Comment:
This field is supposed to be the DSN if connecting via DSN (same as SQLGetInfo(DATA_SOURCE_NAME).

I do not remember if GetDataSourceComponent() returns the DSN, or if it returns the component name that is used for formatting the error message here: https://learn.microsoft.com/en-us/sql/odbc/reference/develop-app/diagnostic-messages?view=sql-server-ver16 . That's supposed to describe the internal component reporting the error.